// cli-highlight-esm/dist/theme.mjs
import { default as _chalk } from 'chalk';
// Always enable colors, even if not auto-detected
const chalk = new _chalk.constructor({ enabled: true });
/**
 * Identity function for tokens that should not be styled (returns the input string as-is).
 * See [[Theme]] for an example.
 */
export const plain = (codePart) => codePart;
/**
 * The default theme. It is possible to override just individual keys.
 */
export const DEFAULT_THEME = {
    /**
     * keyword in a regular Algol-style language
     */
    keyword: chalk.blue,
    /**
     * built-in or library object (constant, class, function)
     */
    built_in: chalk.cyan,
    /**
     * user-defined type in a language with first-class syntactically significant types, like
     * Haskell
     */
    type: chalk.cyan.dim,
    /**
     * special identifier for a built-in value ("true", "false", "null")
     */
    literal: chalk.blue,
    /**
     * number, including units and modifiers, if any.
     */
    number: chalk.green,
    /**
     * literal regular expression
     */
    regexp: chalk.red,
    /**
     * literal string, character
     */
    string: chalk.red,
    /**
     * parsed section inside a literal string
     */
    subst: plain,
    /**
     * symbolic constant, interned string, goto label
     */
    symbol: plain,
    /**
     * class or class-level declaration (interfaces, traits, modules, etc)
     */
    class: chalk.blue,
    /**
     * function or method declaration
     */
    function: chalk.yellow,
    /**
     * name of a class or a function at the place of declaration
     */
    title: plain,
    /**
     * block of function arguments (parameters) at the place of declaration
     */
    params: plain,
    /**
     * comment
     */
    comment: chalk.green,
    /**
     * documentation markup within comments
     */
    doctag: chalk.green,
    /**
     * flags, modifiers, annotations, processing instructions, preprocessor directive, etc
     */
    meta: chalk.grey,
    /**
     * keyword or built-in within meta construct
     */
    'meta-keyword': plain,
    /**
     * string within meta construct
     */
    'meta-string': plain,
    /**
     * heading of a section in a config file, heading in text markup
     */
    section: plain,
    /**
     * XML/HTML tag
     */
    tag: chalk.grey,
    /**
     * name of an XML tag, the first word in an s-expression
     */
    name: chalk.blue,
    /**
     * s-expression name from the language standard library
     */
    'builtin-name': plain,
    /**
     * name of an attribute with no language defined semantics (keys in JSON, setting names in
     * .ini), also sub-attribute within another highlighted object, like XML tag
     */
    attr: chalk.cyan,
    /**
     * name of an attribute followed by a structured value part, like CSS properties
     */
    attribute: plain,
    /**
     * variable in a config or a template file, environment var expansion in a script
     */
    variable: plain,
    /**
     * list item bullet in text markup
     */
    bullet: plain,
    /**
     * code block in text markup
     */
    code: plain,
    /**
     * emphasis in text markup
     */
    emphasis: chalk.italic,
    /**
     * strong emphasis in text markup
     */
    strong: chalk.bold,
    /**
     * mathematical formula in text markup
     */
    formula: plain,
    /**
     * hyperlink in text markup
     */
    link: chalk.underline,
    /**
     * quotation in text markup
     */
    quote: plain,
    /**
     * tag selector in CSS
     */
    'selector-tag': plain,
    /**
     * #id selector in CSS
     */
    'selector-id': plain,
    /**
     * .class selector in CSS
     */
    'selector-class': plain,
    /**
     * [attr] selector in CSS
     */
    'selector-attr': plain,
    /**
     * :pseudo selector in CSS
     */
    'selector-pseudo': plain,
    /**
     * tag of a template language
     */
    'template-tag': plain,
    /**
     * variable in a template language
     */
    'template-variable': plain,
    /**
     * added or changed line in a diff
     */
    addition: chalk.green,
    /**
     * deleted line in a diff
     */
    deletion: chalk.red,
};
/**
 * Converts a [[JsonTheme]] with string values to a [[Theme]] with formatter functions. Used by [[parse]].
 */
export function fromJson(json) {
    const theme = {};
    for (const key of Object.keys(json)) {
        const style = json[key];
        if (Array.isArray(style)) {
            ;
            theme[key] = style.reduce((prev, curr) => (curr === 'plain' ? plain : prev[curr]), chalk);
        }
        else {
            ;
            theme[key] = chalk[style];
        }
    }
    return theme;
}
/**
 * Converts a [[Theme]] with formatter functions to a [[JsonTheme]] with string values. Used by [[stringify]].
 */
export function toJson(theme) {
    const jsonTheme = {};
    for (const key of Object.keys(jsonTheme)) {
        const style = jsonTheme[key];
        jsonTheme[key] = style._styles;
    }
    return jsonTheme;
}
/**
 * Stringifies a [[Theme]] with formatter functions to a JSON string.
 *
 * ```ts
 * import chalk = require('chalk');
 * import {stringify} from 'cli-highlight';
 * import * as fs from 'fs';
 *
 * const myTheme: Theme = {
 *     keyword: chalk.red.bold,
 *     addition: chalk.green,
 *     deletion: chalk.red.strikethrough,
 *     number: plain
 * }
 * const json = stringify(myTheme);
 * fs.writeFile('mytheme.json', json, (err: any) => {
 *     if (err) throw err;
 *     console.log('Theme saved');
 * });
 * ```
 */
export function stringify(theme) {
    return JSON.stringify(toJson(theme));
}
/**
 * Parses a JSON string into a [[Theme]] with formatter functions.
 *
 * ```ts
 * import * as fs from 'fs';
 * import {parse, highlight} from 'cli-highlight';
 *
 * fs.readFile('mytheme.json', 'utf8', (err: any, json: string)  => {
 *     if (err) throw err;
 *     const code = highlight('SELECT * FROM table', {theme: parse(json)});
 *     console.log(code);
 * });
 * ```
 */
export function parse(json) {
    return fromJson(JSON.parse(json));
}
